snowfakery/plugins.py

```diff
@@ -19,17 +16,48 @@ def func1(self, arg1, arg2, arg3=default):
     Despite the name, the Functions namespace could also include
     constants or other namespaces, if that is useful to you.
 
-    A RuntimeContext object is supplied on `self.context` but its
-    interface is not stable yet. Use at your own risk.
+    A RuntimeContext object is supplied on `self.context` with methods
+    context.field_vars() and context.context_vars().
+
+    context.field_vars() are the same field variables that can be used in
+    templates.
+
+    context.context_vars() is a mutable mapping contaiing
+    values that are available to this object template and that of
+    all children.
+
+    context_vars are unique to each plugin and plugins do not see
+    (or corrupt, or overwrite) context_vars for another plugin.
+
+    Plugins can also keep internal state for global data, like any other
+    Python object.
     """
+    def __init__(self, interpreter):
+        self.interpreter = interpreter
+        self.context = PluginContext(self)
 
     def custom_functions(
         self,
-        context: "snowfakery.data_generator_runtime.RuntimeContext",
         *args,
         **kwargs
     ):
-        "Instantiate, contextualize and return a function library"
+        """Instantiate, contextualize and return a function library
+
+        Default behaviour is to return self.Function()."""
         functions = self.Functions()
-        functions.context = context
+        functions.context = self.context
         return functions
+
+
+class PluginContext:
+    "Exposes certain stable internals to plugins"
+
+    def __init__(self, plugin):
+        self.plugin = plugin
+        self.interpreter = plugin.interpreter
+
+    def field_vars(self):
+        return self.interpreter.current_context.field_vars()
```

Comment:
Is it desired or not that this now includes the field_funcs?

Reply:
Yes I think its better to be consistent. Now plugins can use the field_funcs in case they are useful.

Reply:
I think that the only reason they were ever separate is because I had to do some curry magic with the field_funcs.